src/FileType_XDMF.f90

```diff
@@ -0,0 +1,1236 @@
+!++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++!
+!                          Futility Development Group                          !
+!                             All rights reserved.                             !
+!                                                                              !
+! Futility is a jointly-maintained, open-source project between the University !
+! of Michigan and Oak Ridge National Laboratory.  The copyright and license    !
+! can be found in LICENSE.txt in the head directory of this repository.        !
+!++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++!
+!> @brief Module defines objects for representing an XDMF file as a hierarchical
+!> mesh
+!>
+!> This module reads an XDMF file and stores the information in a hierarchical
+!> mesh type. It can also write the hierarchical mesh to XDMF.
+!++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++!
+MODULE FileType_XDMF
+#include "Futility_DBC.h"
+USE ExceptionHandler
+USE Futility_DBC
+USE ISO_FORTRAN_ENV
+USE IntrType
+USE Strings
+USE FileType_Base
+USE FileType_XML
+USE FileType_HDF5
+USE ParameterLists
+
+IMPLICIT NONE
+PRIVATE
+
+#ifdef FUTILITY_HAVE_HDF5
+! Public members
+PUBLIC :: XDMFFileType
+PUBLIC :: XDMFMeshType
+PUBLIC :: XDMFTopologyList
+PUBLIC :: ASSIGNMENT(=)
+
```

Comment:
I see a lot of indexing with 1, 2, 3, 4 integers which can be confusing when newly looking at the code. Can you assign named integers here so that more context is given for the indexing? I.e. GEOMETRY_IDX=1, TOPOLOGY_IDX=2, etc

Reply:
Added where possible. A lot of the 1 indexing is due to the assumption that there is only one child. There are checks for this, but now that I have it as error handling it should be more clear that the assumption exists.